### DISCUSSION TITLE
Unexpected onDelete called just after Get.toNamed(), controller deleted.

### DISCUSSION THREAD

**Discussion by techieasif** [jonataslaw/getx#1360] [Q&A]
I'm just navigating to new page using named routes, as soon as toNamed trigged new screen flashed, then closed console shows 'onDelete called', REPLACE ROUTE with navigating one.

1. main.dart
```
class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return GetMaterialApp(
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.green,
      ),
      initialRoute: AppRoutes.splash,
      getPages: Pages.getPages,
    );
  }
}
```

2. Pages
```
abstract class Pages {
  static List<GetPage> getPages = [
    GetPage(name: AppRoutes.splash, page: () => SplashPage(), binding: SplashBinding()),
    GetPage(name: AppRoutes.login, page: () => LoginPage(), binding: LoginBinding()),
    GetPage(name: AppRoutes.home, page: () => HomePage(), binding: HomeBinding(), transition: Transition.fadeIn),
    //issue in last one.
    GetPage(name: AppRoutes.newTask, page: () => NewTaskPage(), binding: TaskBindings(), transition: Transition.fadeIn),
  ];
}
````
NOTE:  Navigating to NewTaskPage(), 

3. Binding
```
class TaskBindings extends Bindings{
  @override
  void dependencies() {
   Get.put<TaskController>(TaskController());
  }
}
````

4. task controller

```
class TaskController extends GetxController{

@override
  void onInit() {
    //todo: fetch Task info.
    super.onInit();
  }

}
```

5. Task Page 

```
class NewTaskPage extends GetView<TaskController> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.red,
      appBar: AppBar(
        title: Text("NEW TASK"),
      ),
      body: Center(child: Text("No Working"),),
    );
  }
}
```

6. Navigating function

```
void navigate2NewTask() {
    Get.toNamed(AppRoutes.newTask);
  }
```
7. calling above function
```
class HomePage extends GetView<HomeController> {
  final GlobalKey<ScaffoldState> _homeScfKey = GlobalKey<ScaffoldState>();
  @override
  Widget build(BuildContext context) {
    return Scaffold(
        floatingActionButton: FloatingActionButton(
//SEE HERE
          onPressed: () => controller.navigate2NewTask(),
          child: Icon(Icons.add),
        ),
        key: _homeScfKey,
        drawer: HomeDrawer(),
)
}
}
```

New task page flashed and then the console shows this:
```
[GETX] GOING TO ROUTE /new-task
[GETX] Instance "TaskController" has been created
[GETX] Instance "TaskController" has been initialized
flutter: Splash navigation
[GETX] REPLACE ROUTE /new-task
[GETX] NEW ROUTE /app-home
flutter: Splash navigation
[GETX] "TaskController" onDelete() called
[GETX] "TaskController" deleted from memory
```

Please help.
@jasonlaw

**Comment by flutterwithfaiz** [comment]
I am facing the same problem. Is there any solution available?

**Comment by techieasif** [comment]
Use Flutter bloc/cubit.


On Tue, 10 Aug 2021 at 16:14, flutterwithfaiz ***@***.***>
wrote:

> I am facing the same problem. Is there any solution available?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jonataslaw/getx/discussions/1360#discussioncomment-1153631>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGFYZXXGBGSTISUDLWBLJKTT4D7ILANCNFSM43P6IJ2A>
> .
>


-- 
With Warm Regards,

*MOHD ASIF*,
LinkedIn: https://www.linkedin.com/in/techieasif

**Comment by HamzaMihfad** [comment]
You should call Get.find on the very first screen on your app, so GetX will not delete it:
`YourController yourController = Get.find<YourController>(); `
it worked for me ✅

**Reply by devang5114** [reply]
same bro

**Comment by abdalwhabSalahaldeen** [comment]
try this 
AddPointController yourController =
      Get.put(AddPointController(), permanent: true); 
its work with me 

 permanent: true 
this will stop remove data

**Comment by reasje** [comment]
Hi, I had the same bug but in my case, I used `Navigator.pushReplacement()` for navigation so GetX didn't know anything about the route I am in and delete the controller for some reason. 
Summary: Do not use Navigator for navigation when using GetX controller.

**Reply by niamulhasan** [reply]
This is the case, I found when `back gesture` or `back button` is pressed by default flutter calls `Navigator.pop()` is invoked, then this issue is created. But when `Get.back()` is used it behaves as intended.

**Comment by NegusNati** [comment]
you first need to persist the controller by calling it on the main method like

`    Get.find<PopularProductController>()
        .initProduct(Get.find<CartController>()); `      
then when ever you initialize your dependencies use permanent: True or fenix: True for `Get.lazyPut()` . 
```
Get.lazyPut(
       fenix: true,
      () => CartController(cartRepo: CartRepo()));
//or
     Get.put( CartController(cartRepo: CartRepo()), permanent:true);
```
